src/lib.rs

```diff
@@ -0,0 +1,631 @@
+                KnownEncoding::AppCustom,
+                payload.format.to_string().into(),
+            ))
+            .priority(priority)
+            .with_attachment(attachment.build());
+
+        // Send data
+        putbuilder
+            .res()
+            .await
+            .map_err(|_| UStatus::fail_with_code(UCode::Internal, "Unable to send with Zenoh"))?;
+
+        Ok(())
+    }
+
+    async fn send_response(
+        &self,
+        zenoh_key: &str,
+        payload: UPayload,
+        attributes: UAttributes,
+    ) -> Result<(), UStatus> {
+        // Get the data from UPayload
+        let Some(Data::Value(buf)) = payload.data else {
+            // TODO: Assume we only have Value here, no reference for shared memory
+            return Err(UStatus::fail_with_code(
+                UCode::InvalidArgument,
+                "Invalid data",
+            ));
+        };
+
+        // Serialized UAttributes into protobuf
+        let mut attr = vec![];
+        let Ok(()) = attributes.encode(&mut attr) else {
+            return Err(UStatus::fail_with_code(
+                UCode::InvalidArgument,
+                "Unable to encode UAttributes",
+            ));
+        };
+        // Get reqid
+        let reqid = ULinkZenoh::uuid_to_string(&attributes.reqid.ok_or(
+            UStatus::fail_with_code(UCode::InvalidArgument, "reqid doesn't exist"),
+        )?);
+
+        // Add attachment and payload
+        let mut attachment = AttachmentBuilder::new();
+        attachment.insert("uattributes", attr.as_slice());
+        // Send back query
+        let value = Value::new(buf.into()).encoding(Encoding::WithSuffix(
+            KnownEncoding::AppCustom,
+            payload.format.to_string().into(),
+        ));
+        let reply = Ok(Sample::new(
+            KeyExpr::new(zenoh_key.to_string()).map_err(|_| {
+                UStatus::fail_with_code(UCode::Internal, "Unable to create Zenoh key")
+            })?,
+            value,
+        ));
+        let query = self
+            .query_map
+            .lock()
+            .unwrap()
+            .get(&reqid)
+            .ok_or(UStatus::fail_with_code(
+                UCode::Internal,
+                "query doesn't exist",
+            ))?
+            .clone();
+
+        // Send data
+        // TODO: Unable to use unwrap in with_attachment (Attachment doesn't have Debug trait)
+        query
+            .reply(reply)
+            .with_attachment(attachment.build())
+            .map_err(|_| UStatus::fail_with_code(UCode::Internal, "Unable to add attachment"))?
+            .res()
+            .await
+            .map_err(|_| UStatus::fail_with_code(UCode::Internal, "Unable to reply with Zenoh"))?;
+
+        Ok(())
+    }
+}
+
+#[async_trait]
+impl RpcClient for ULinkZenoh {
+    async fn invoke_method(
+        &self,
+        topic: UUri,
+        payload: UPayload,
+        attributes: UAttributes,
+    ) -> RpcClientResult {
```

Comment:
One thing I ran into when writing a uStreamer implementation is that the RpcServer should be responsible for providing (at a minimum) UAttributes back and RpcClientResult should return in addition to a UPayload, at a minimum the UAttributes. Why? This allows for a uStreamer implementation to properly route a response back over another transport. Let's discuss more in our meeting tomorrow.

Reply:
There's a [spec update](https://github.com/eclipse-uprotocol/up-spec/issues/61) which has invoke_method return a UMessage instead. We need to update up-rust's invoke_method signature before you could make that change tho.